design-decisions/2018-09-17-Builtin-Map.md

```diff
@@ -0,0 +1,71 @@
+Builtin Maps
+============
+
+Background
+----------
+
+K provides a builtin `Map` sort representing partial maps on sort `K`.
+Here "partial" means that some keys may not be associated with any value.
+KEVM relies on concrete operations only, but we will eventually need to support
+symbolic operations.
+
+Operations
+----------
+
+The builtin `Map` domain should support operations equivalent to:
+
+```
+module MAP
+    sort Map{} []
+
+    // unit: The empty map (no keys).
+    symbol unit{}() : Map{} []
+
+    // element(key, value): A map from one key to one value.
+    symbol element{}(K{}, K{}) : Map{} []
+
+    // concat(map1, map2): Combine both maps so that a key is in the result if
+    // it is in either operand.
+    symbol concat{}(Map{}, Map{}) : Map{} []
+
+    // update(map, key, value): Update the map so that key refers to value.
+    symbol update{}(Map{}, K{}, K{}) : Map{} []
+
+    // lookup(map, key): Lookup key in map and return the associated value.
+    // Return \bottom if the key is missing.
+    symbol lookup{}(Map{}, K{}) : K{} []
+
+    // inKeys(map, key): Is key in map?
+    symbol inKeys{}(Map{}, K{}) : Bool{} []
+endmodule
+```
+
+Questions:
+
+- What does `update` do if the key is missing from the map?
```

Comment:
@ttuegel @traiansf I think the `update(m,k,v)` should add the new entry `k |-> v` when `k` doesn't exist in `m`. In that sense, we may want to change the name to `insert` to capture this behavior.

What do you think?

Reply:
I agree that it makes sense for `update` to insert a missing key, but I don't want to change the name. If it had been named `insert` instead, I would have asked, what happens if the key is already in the map? :smile: